### PR TITLE
Adds the Shimmer Component, and gives gold datum items a fresh coat of paint.

### PR DESCRIPTION
**Pull request by ArcaneMusic**
## About The Pull Request
image
Animated version: https://cdn.discordapp.com/attachments/326831214667235328/632460559035400223/2019-10-12_02-11-57.mp4

Datum gold items right now were looking a bit bland, so this adds a new decal component, similar to the blood overlay component, called shimmer. It's applied to all items with gold as a material, making them look shiny and more attractive than before.

## Why It's Good For The Game

Well, the original intent was to apply this mainly to datum objects, like gold datum toolboxes, knights armor, etc, but it looks alright on most of the items I've seen that contain enough gold to apply on it. This'll help differentiate between bananium datum objects and gold, without needing to try and find a more or less gold shade of yellow/orange. This also means that this component could be added to anything else meant to look expensive and valuable.

## Changelog
:cl:
tweak: Gold looks a bit more expensive and shiny now.
imageadd: added a new sprite for gold datum materials, overlaid over the original sprites.
/:cl:

So fair warning, the one thing I still haven't ironed out is a fairly outstanding old big regarding how decal overlays are done, which is best seen on metal rods. If you apply the blood component on something like rods, it adds the whole blood overlay to the single rod sprite, and the same happens on objects like gold bars with stacking qualities. I'll see if I can fix it but it's a thing since about when the blood decal component was added in, so we'll see how that works out.

### CODE REVIEW

**Comment by PKPenguin321** [comment]
Would be nice if it were actually animated

**Comment by nemvar** [comment]
Yeah, the lack of wave filters makes me very sad.

**Comment by Qustinnus** [comment]
I agree with Nemvar. Can you post a video of this on a 2x4 gold table setup?

File: code/game/objects/items/stacks/stack.dm [hunk 1]
```diff
@@ -89,6 +93,14 @@
...
 		icon_state = "[initial(icon_state)]_3"
+
+	if(GetComponent(overlay))
+		if(src.GetComponent(/datum/component/decal/shimmer))
+			S.TakeComponent(src)
+			src.AddComponent(/datum/component/decal/shimmer)
+		if(src.GetComponent(/datum/component/decal/blood))
+			B.TakeComponent(src)
+			src.AddComponent(/datum/component/decal/blood)
 	..()
```
Comment (Qustinnus) [inline code/game/objects/items/stacks/stack.dm]: There is surely a better way of doing this? This syntax looks kind of odd.

**Comment by ishitbyabullet** [comment]
Tone down the shimmering. This is mostly noticeable on things like the gold trophy cup where it looks more like a line is scanning through it repeatedly.

**Comment by ArcaneMusic** [comment]
> Would be nice if it were actually animated

@PKPenguin321 It is animated, it just didn't show well on the video I linked.



> Yeah, the lack of wave filters makes me very sad.
> 
> I don't expect you to add them but it feels like the animation should be faster and also not happen that often so it has more of that bling bling feel, you dig?

I can certainly change the animation to make it closer to that, sure.



> I agree with Nemvar. Can you post a video of this on a 2x4 gold table setup?

So, apparently this doesn't apply to structures like gold walls or floors, just objects so far. I'll see what I can do to work that in.

File: code/game/objects/items/stacks/stack.dm (hunk 1 above)
Comment (ArcaneMusic) [reply code/game/objects/items/stacks/stack.dm]: Almost certainly there is, this is trying to fix that stacked items overlay thing, but I haven't had any luck with a fix so far, so this'll likely disappear once I get that worked out.

File: code/datums/components/decals/materials.dm [hunk 2]
```diff
@@ -0,0 +1,39 @@
+/datum/component/decal/gold
+	dupe_mode = COMPONENT_DUPE_UNIQUE
+
+/datum/component/decal/gold/Initialize(_icon, _icon_state, _dir, _cleanable=CLEAN_STRENGTH_FIBERS, _color, _layer=ABOVE_OBJ_LAYER)
+	if(!isitem(parent))
```
Comment (Qustinnus) [inline code/datums/components/decals/materials.dm]: @ArcaneMusic this is probably why kek

**Comment by ArcaneMusic** [comment]
Alright, so, as an unintended consequence of how decals work, one thing that was causing fairly jarring graphical issues was that then decals were applied, (very noticeably in game, by the blood overlay), applying a decal component would apply the blood that would otherwise be applied to the initial icon_state. This looks a LOT better on most items that have overlays or variable appearance, but for the purposes of applying the shimmer overlay I was running into the same problem, which was especially noticeable on stacks of gold bars, so I've _HOPEFULLY_ fixed both the big with blood overlays and the issue with shimmer overlays!
image
As requested, here's 4x4 gold tables. There's definitely something I'll have to fix involving icon smoothing there, but this is certainly a good start for now.

File: code/datums/components/decal.dm [hunk 3]
```diff
@@ -13,7 +13,6 @@
 	first_dir = _dir
 	description = _description
 	cleanable = _cleanable
-
```
Comment (ninjanomnom) [inline code/datums/components/decal.dm]: ?

File: code/datums/components/decals/blood.dm [hunk 4]
```diff
@@ -9,12 +9,15 @@
...
 /datum/component/decal/blood/generate_appearance(_icon, _icon_state, _dir, _layer, _color)
 	var/obj/item/I = parent
+	I.update_icon()
 	if(!_icon)
 		_icon = 'icons/effects/blood.dmi'
 	if(!_icon_state)
 		_icon_state = "itemblood"
 	var/icon = initial(I.icon)
-	var/icon_state = initial(I.icon_state)
+	var/icon_state = I.icon_state
```
Comment (ninjanomnom) [inline code/datums/components/decals/blood.dm]: If you're going to make this support using current icon state then you need to update the decal whenever the icon changes. You'll want a signal in update_icon()

File: code/datums/components/decals/shimmer.dm [hunk 5]
```diff
@@ -0,0 +1,48 @@
+/datum/component/decal/shimmer
+	dupe_mode = COMPONENT_DUPE_UNIQUE
+
+/datum/component/decal/shimmer/Initialize(_icon, _icon_state, _dir, _cleanable=CLEAN_STRENGTH_FIBERS, _color, _layer=ABOVE_OBJ_LAYER)
+	. = ..()
+	RegisterSignal(parent, COMSIG_ATOM_GET_EXAMINE_NAME, .proc/get_examine_name)
+
+/datum/component/decal/shimmer/generate_appearance(_icon, _icon_state, _dir, _layer, _color)
```
Comment (ninjanomnom) [inline code/datums/components/decals/shimmer.dm]: Just copy pasting this whole proc is awful

**Review by ninjanomnom — CHANGES_REQUESTED** [review]

**Comment by ninjanomnom** [comment]
I'm really not a fan of how this is all accomplished and I don't think the load on the server for all this is worth the change. I do like the look though.

If you really want this I'd wait for 513 when this should become far easier to accomplish with similar effects with render targets.

**Comment by ninjanomnom** [comment]
Other maintainers can feel free to disagree but I don't think this feature is worth the cost in maintenance and server resources with how it's implemented here, there are other options but they require some more extensive refactors and deserve their own pr.

**Comment by ArcaneMusic** [comment]
That's fair, I'll rework this once 513 comes around.

**Pull request closed without merging**